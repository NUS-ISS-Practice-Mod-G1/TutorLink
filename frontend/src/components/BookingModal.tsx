import React, { useState, useEffect } from "react";

export interface BookingModalProps {
  lessonTypes: string[];
  slot: { date: Date; slot: { start: string; end: string } };
  hourlyRate?: number; // ‚úÖ new prop
  onClose: () => void;
  onConfirm: (lessonType: string) => void | Promise<void>;
}

const BookingModal: React.FC<BookingModalProps> = ({
  lessonTypes,
  slot,
  hourlyRate,
  onClose,
  onConfirm,
}) => {
  const [selectedLessonType, setSelectedLessonType] = useState<string>(lessonTypes[0] || "");
  const [estimatedCost, setEstimatedCost] = useState<number | null>(null);

  useEffect(() => {
    if (hourlyRate && slot.slot.start && slot.slot.end) {
      // Calculate hours
      const startTime = new Date(`1970-01-01T${slot.slot.start}:00`);
      const endTime = new Date(`1970-01-01T${slot.slot.end}:00`);
      const diffHours = (endTime.getTime() - startTime.getTime()) / (1000 * 60 * 60);
      if (diffHours > 0) {
        setEstimatedCost(diffHours * hourlyRate);
      } else {
        setEstimatedCost(null);
      }
    }
  }, [hourlyRate, slot]);

  const handleConfirm = () => {
    if (!selectedLessonType) return;
    onConfirm(selectedLessonType);
  };

  return (
    <div className="fixed inset-0 flex items-center justify-center bg-black bg-opacity-40 z-50">
      <div className="bg-white rounded-xl shadow-lg p-6 w-96">
        <h2 className="text-lg font-semibold mb-4">Confirm Booking</h2>

        <p className="mb-2">
          üìÖ {slot.date.toDateString()} | ‚è∞ {slot.slot.start} - {slot.slot.end}
        </p>

        {/* Lesson Type Selector */}
        <label className="block mb-2 text-sm font-medium">Lesson Type</label>
        <select
          className="w-full border rounded-md p-2 mb-4"
          value={selectedLessonType}
          onChange={(e) => setSelectedLessonType(e.target.value)}>
          {lessonTypes.map((lt, idx) => (
            <option key={idx} value={lt}>
              {lt}
            </option>
          ))}
        </select>

        {/* Optional estimated cost */}
        {hourlyRate && estimatedCost !== null && (
          <div className="mb-4 p-3 bg-blue-50 rounded-md text-sm text-gray-700">
            <p>
              üí∞ Estimated Cost:{" "}
              <span className="font-bold text-blue-600">SGD {estimatedCost.toFixed(2)}</span>
            </p>
            <p className="text-xs text-gray-500">({hourlyRate} SGD/hr √ó duration)</p>
          </div>
        )}

        <div className="flex justify-end gap-2">
          <button onClick={onClose} className="px-4 py-2 rounded-md bg-gray-200 hover:bg-gray-300">
            Cancel
          </button>
          <button
            onClick={handleConfirm}
            className="px-4 py-2 rounded-md bg-blue-600 text-white hover:bg-blue-700">
            Confirm
          </button>
        </div>
      </div>
    </div>
  );
};

export default BookingModal;
