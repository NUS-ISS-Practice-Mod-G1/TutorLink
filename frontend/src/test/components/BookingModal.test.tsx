import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import BookingModal, { BookingModalProps } from "@/components/BookingModal";

describe("BookingModal", () => {
  const mockOnClose = jest.fn();
  const mockOnConfirm = jest.fn();

  const baseProps: BookingModalProps = {
    lessonTypes: ["Math", "Science", "English"],
    slot: { date: new Date("2025-11-10"), slot: { start: "10:00", end: "12:00" } },
    hourlyRate: 50,
    onClose: mockOnClose,
    onConfirm: mockOnConfirm,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  // ---------- BASIC RENDER ----------
  it("renders lesson type dropdown and date/time info", () => {
    render(<BookingModal {...baseProps} />);
    expect(screen.getByText("Confirm Booking")).toBeInTheDocument();
    expect(screen.getByText("Math")).toBeInTheDocument();
    expect(screen.getByText("Please confirm your booking details below:")).toBeInTheDocument();
  });

  // ---------- COST CALCULATION ----------
  it("calculates estimated cost correctly (2 hours × 50 SGD/hr)", async () => {
    render(<BookingModal {...baseProps} />);
    expect(await screen.findByText("SGD 100.00")).toBeInTheDocument();
    expect(screen.getByText("(50 SGD/hr × 2 hour(s))")).toBeInTheDocument();
  });

  // ---------- OVERNIGHT SLOT ----------
  it("handles overnight slot correctly when end < start (23:00 → 01:00)", async () => {
    render(
      <BookingModal
        {...baseProps}
        slot={{ date: new Date("2025-11-10"), slot: { start: "23:00", end: "01:00" } }}
      />
    );

    // end time is next day, diff = 2h
    await waitFor(() => {
      expect(screen.getByText("SGD 100.00")).toBeInTheDocument();
    });
  });

  // ---------- ZERO-DURATION / INVALID SLOT ----------
  it("does not show estimated cost when duration ≤ 0", async () => {
    const baseProps: BookingModalProps = {
      lessonTypes: ["Math", "Science", "English"],
      slot: { date: new Date("2025-11-10"), slot: { start: "10:00", end: "10:00" } },
      hourlyRate: 50,
      onClose: mockOnClose,
      onConfirm: mockOnConfirm,
    };
    render(
      <BookingModal
        {...baseProps}
      />
    );

    await waitFor(() => {
      expect(screen.queryByText(/Estimated Cost/)).not.toBeInTheDocument();
    });
  });

  // ---------- MISSING HOURLY RATE ----------
  it("does not render estimated cost when hourlyRate is undefined", () => {
    render(<BookingModal {...baseProps} hourlyRate={undefined} />);
    expect(screen.queryByText(/Estimated Cost/)).not.toBeInTheDocument();
  });

  // ---------- LESSON TYPE CHANGE ----------
  it("updates selected lesson type when user changes selection", () => {
    render(<BookingModal {...baseProps} />);
    const select = screen.getByDisplayValue("Math");
    fireEvent.change(select, { target: { value: "English" } });
    expect(screen.getByDisplayValue("English")).toBeInTheDocument();
  });

  // ---------- CONFIRM ----------
  it("calls onConfirm with the selected lesson type", () => {
    render(<BookingModal {...baseProps} />);
    fireEvent.click(screen.getByText("Confirm"));
    expect(mockOnConfirm).toHaveBeenCalledWith("Math");
  });

  // ---------- CANCEL ----------
  it("calls onClose when Cancel button is clicked", () => {
    render(<BookingModal {...baseProps} />);
    fireEvent.click(screen.getByText("Cancel"));
    expect(mockOnClose).toHaveBeenCalled();
  });

  // ---------- EMPTY LESSON LIST ----------
  it("does not call onConfirm if no lesson types exist", () => {
    render(<BookingModal {...baseProps} lessonTypes={[]} />);
    fireEvent.click(screen.getByText("Confirm"));
    expect(mockOnConfirm).not.toHaveBeenCalled();
  });

  // ---------- SNAPSHOT ----------
  it("matches snapshot", async () => {
    const { container } = render(<BookingModal {...baseProps} />);
    await waitFor(() => screen.findByText("SGD 100.00"));
    expect(container).toMatchSnapshot();
  });
});
