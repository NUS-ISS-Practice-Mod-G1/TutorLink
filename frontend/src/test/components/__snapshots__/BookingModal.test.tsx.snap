// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`BookingModal matches snapshot 1`] = `
<div>
  <div
    class="fixed inset-0 flex items-center justify-center bg-black bg-opacity-40 z-50"
  >
    <div
      class="bg-white rounded-xl shadow-lg p-6 w-96"
    >
      <h2
        class="text-lg font-semibold mb-4"
      >
        Confirm Booking
      </h2>
      <p
        class="mb-4"
      >
        Please confirm your booking details below:
      </p>
      <p
        class="mb-2"
      >
        üìÖ 
        <strong>
          Date:
        </strong>
         
        Mon Nov 10 2025
         
        <br />
        ‚è∞ 
        <strong>
          Time:
        </strong>
         
        10:00
         - 
        12:00
      </p>
      <label
        class="block mb-2 text-sm font-medium"
      >
        Choose Lesson Type:
      </label>
      <select
        class="w-full border rounded-md p-2 mb-4"
      >
        <option
          value="Math"
        >
          Math
        </option>
        <option
          value="Science"
        >
          Science
        </option>
        <option
          value="English"
        >
          English
        </option>
      </select>
      <div
        class="mb-4 p-3 bg-blue-50 rounded-md text-sm text-gray-700"
      >
        <p>
          üí∞ Estimated Cost:
           
          <span
            class="font-bold text-blue-600"
          >
            SGD 
            100.00
          </span>
        </p>
        <p
          class="text-xs text-gray-500"
        >
          (
          50
           SGD/hr √ó 
          2
           hour(s))
        </p>
      </div>
      <div
        class="flex justify-end gap-2"
      >
        <button
          class="px-4 py-2 rounded-md bg-gray-200 hover:bg-gray-300"
        >
          Cancel
        </button>
        <button
          class="px-4 py-2 rounded-md bg-blue-600 text-white hover:bg-blue-700"
        >
          Confirm
        </button>
      </div>
    </div>
  </div>
</div>
`;
